volumes/vds/py/vds_functions.py

```diff
@@ -0,0 +1,352 @@
+import logging
+
+LOGGER = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+def pull_raw_vdsdata(rds_conn, itsc_conn, start_date):
+#pulls data from ITSC table `vdsdata` and inserts into RDS table `vds.raw_vdsdata`
+
+    import pandas as pd
+    from numpy import nan
+    from psycopg2 import sql, Error
+    from psycopg2.extras import execute_values
+
+    # Pull raw data from Postgres database
+    raw_sql = sql.SQL('''SELECT 
+        d.divisionid,
+        d.vdsid,
+        d.timestamputc, --timestamp in INTEGER (UTC)
+        d.lanedata
+    FROM public.vdsdata AS d
+    WHERE
+        timestamputc >= extract(epoch from timestamp with time zone {start})
+        AND timestamputc < extract(epoch from timestamp with time zone {start} + INTERVAL '1 DAY')
+        AND d.divisionid = 2; --other is 8001 which are traffic signal detectors and are mostly empty.
+    ''').format(
+        start = sql.Literal(start_date + ' 00:00:00 EST5EDT')
+    )
+
+    try: 
+        with itsc_conn.get_conn() as con:
+            raw_data = pd.read_sql(raw_sql, con)
+            LOGGER.info('Fetching vdsdata')
+    except Error as exc:
+        LOGGER.critical('Error fetching vdsdata.')
+        LOGGER.critical(exc)
+        con.close()
+    
+    # Transform raw data
+    transformed_data = transform_raw_data(raw_data)
+    transformed_data = transformed_data.replace(nan, None)
+    transformed_data = transformed_data[[x is not None for x in transformed_data['lane']]] #remove lane is None
+    data_tuples = [tuple(x) for x in transformed_data.to_numpy()]
+
+    try:
+        with rds_conn.get_conn() as con:
+            with con.cursor() as cur:
+                # Drop records for the current date
+                #drop_query = sql.SQL(f"DELETE FROM vds.raw_vdsdata WHERE datetime_bin::date = '{dt}'")
+                #con.execute(drop_query)
```

Comment:
Why is this query commented?

Reply:
It depends on how the DAG works. Some DAGs might be updating the entire dataset every time they run. Hence, no need to delete old/incomplete data. Some other DAGs require us to delete this data manually before backfilling (like miovision). But I think your approach here is more robust.